### PR TITLE
feat(chart-heatfield): Chart Heatfield Overlay should have a close bu…

### LINKED ISSUE #732: Chart Heatfield: Overlay should have a close button
# Feature Request

To increase consistent behavior with our overlays we decided to add a close button to the chart-heatfield overlay. 

The close button should be a secondary button and the focus should be trapped inside the overlay. Similar to what is done for the selection area overlay or the context dialog overlay.

For a longer discussion see #725

### PR DESCRIPTION
**Pull request by Sherif-Elhefnawy**
### <strong>Pull Request</strong>

Chart Heatfield Overlay should have a close button

- Close button has been added
- Added tests

#### Type of PR

Feature (non-breaking change which adds functionality)

Fixes #732

### CODE REVIEW

**Comment by Sherif-Elhefnawy** [comment]
Thanks, will take a look at it again

File: libs/barista-components/chart/src/heatfield/chart-heatfield.ts [hunk 1]
```diff
@@ -223,6 +223,14 @@ export class DtChartHeatfield extends _DtHeatfieldMixinBase
...
   }
 
+  /**
+   * Closes the context dialog
+   * @internal
+   */
+  close(): void {
+    this.active = false;
+  }
+
```
Comment (ffriedl89) [inline libs/barista-components/chart/src/heatfield/chart-heatfield.ts:233]: This method is doing the same thing as setting active to false on the instance. I'd rather keep the API surface small and not add this method to do the same thing

File: libs/barista-components/chart/src/heatfield/chart-heatfield.spec.ts [hunk 2]
```diff
@@ -189,6 +197,27 @@ describe('DtChartHeatfield', () => {
...
+      it('should include an x button', () => {
+        expect(closeOverlayButton).not.toBeNull();
+      });
+
+      it('clicking on x button should close overlay', () => {
+        closeOverlayButton.click();
+        fixture.detectChanges();
+        expect(marker.classList).not.toContain('dt-chart-heatfield-active');
+      });
+    });
```
Comment (ffriedl89) [inline libs/barista-components/chart/src/heatfield/chart-heatfield.spec.ts:219]: It would be awesome if you could add a testcase for the escape behavior

Comment (Sherif-Elhefnawy) [reply libs/barista-components/chart/src/heatfield/chart-heatfield.spec.ts:219]: Already thought about it after lukas's comment but let's fix the trap focus first :)

File: libs/barista-components/chart/src/heatfield/chart-heatfield.ts (hunk 1 above)
Comment (Sherif-Elhefnawy) [reply libs/barista-components/chart/src/heatfield/chart-heatfield.ts:233]: Alright, valid point

**Comment by ffriedl89** [comment]
A stylelint check also failed in the ci you can check it out by clicking the details link in the check :)

**Comment by Sherif-Elhefnawy** [comment]
Just for future reference: 

cdkTrapFocus actually works fine, the problem lies if you have div or span marked with cdkTrapFocus that doesn't have focusable elements, then the trapfocus doesn't work and "escapes".

this doesn't work:

```
<div cdkTrapFocus>
<a class='dt-link'>Problem</a>
</div>
```
this works:

```
<div cdkTrapFocus>
<a href='#' class='dt-link'>Problem</a>
</div>
```

File: libs/barista-components/chart/src/heatfield/chart-heatfield.ts (hunk 1 above)
Comment (Sherif-Elhefnawy) [reply libs/barista-components/chart/src/heatfield/chart-heatfield.ts]: done

**Comment by ffriedl89** [comment]
@Sherif-Elhefnawy I've checked out the PR - code looks good so far.
I've noticed that pressing enter causes weird behaviour. See the gif
heatfield-816

Could you please take a look? 
After you've fixed this issue it would be awesome if you could squash your commits into one feat commit so we have a proper changelog afterwards. 

Thx for helping out!

File: tools/schematics/package-lock.json [hunk 3]
```diff
@@ -0,0 +1,1907 @@
+{
```
Comment (tomheller) [inline tools/schematics/package-lock.json:1]: Can you please remove this file from the commit. 
I guess schematics should not be installed within the source of the tools folder.

File: .gitignore [hunk 4]
```diff
@@ -9,6 +9,7 @@
 # dependencies
 /node_modules
 yarn.lock
+/tools/schematics/node_modules/
```
Comment (tomheller) [inline .gitignore:12]: The schematics should not be installed within the tools folder, I guess. 
Can you please remove this line?

File: libs/barista-components/chart/src/heatfield/chart-heatfield.ts [hunk 5]
```diff
@@ -122,6 +122,7 @@ export class DtChartHeatfield extends _DtHeatfieldMixinBase
     return this._active;
   }
   set active(val: boolean) {
+    console.log('active', val);
```
Comment (tomheller) [inline libs/barista-components/chart/src/heatfield/chart-heatfield.ts:125]:
```suggestion
```
@ffriedl89 you left some changes in on friday, could you remove them?

File: libs/barista-components/chart/src/heatfield/chart-heatfield.ts [hunk 6]
```diff
@@ -218,8 +219,7 @@ export class DtChartHeatfield extends _DtHeatfieldMixinBase
    * @internal
    */
   _toggleActive(): void {
-    this._overlayAnimationState =
-      this._overlayAnimationState === 'fadeIn' ? 'void' : 'fadeIn';
+    console.log('active toggled');
```
Comment (tomheller) [inline libs/barista-components/chart/src/heatfield/chart-heatfield.ts:222]:
```suggestion
```
@ffriedl89 you left some changes in on friday, could you remove them?

File: libs/barista-components/chart/src/heatfield/chart-heatfield.ts [hunk 7]
```diff
@@ -228,8 +228,11 @@ export class DtChartHeatfield extends _DtHeatfieldMixinBase
    * @internal
    */
   _handleKeydown(event: KeyboardEvent): void {
-    if (_readKeyCode(event) === ENTER) {
-      this._toggleActive();
+    if (_readKeyCode(event) === ESCAPE) {
+      this.active = false;
+    }
```
Comment (tomheller) [inline libs/barista-components/chart/src/heatfield/chart-heatfield.ts:233]:
```suggestion
```
I think once would be enough, @ffriedl89 I guess this happend with your rebase conflict, maybe you can take care of that one as well.

File: .gitignore (hunk 4 above)
Comment (Sherif-Elhefnawy) [reply .gitignore:12]: Was it something wrong on my side that caused it to get installed there?

**Comment by ffriedl89** [comment]
I'll take care of the mess i made - sorry @Sherif-Elhefnawy

**Comment by Sherif-Elhefnawy** [comment]
> 
> 
> I'll take care of the mess i made - sorry @Sherif-Elhefnawy

No problem, thanks @ffriedl89

**Comment by ffriedl89** [comment]
Everything cleaned up

File: libs/barista-components/chart/src/heatfield/chart-heatfield.html [hunk 8]
```diff
@@ -4,8 +4,7 @@
   [style.visibility]="_isValidStartEndRange ? 'visible' : 'hidden'"
   #marker="cdkOverlayOrigin"
   cdk-overlay-origin
-  (click)="_toggleActive()"
-  (keydown)="_handleKeydown($event)"
+  (click)="_toggleActive($event)"
```
Comment (Sherif-Elhefnawy) [inline libs/barista-components/chart/src/heatfield/chart-heatfield.html:7]: One thing that might have slipped, "toggleActive" doesn't have a parameter

File: .gitignore (hunk 4 above)
Comment (tomheller) [reply .gitignore]: I don't really know why this was installed here, tbh. 
Thank you for cleaning it up though.

Comment (Sherif-Elhefnawy) [reply .gitignore]: Might be the intellij npm install then, It was fabian who removed it in the end :)

**Comment by ffriedl89** [comment]
Cerry picked into 6.x

**Pull request merged**